### PR TITLE
Move watcher to use seq# and primary term for concurrency control

### PR DESCRIPTION
**Pull request by bleskes**
Watcher uses internal versioning to make sure a watch isn't changed when updating a status. This PR moves that logic to sequence numbers and opens up the option for users to also use seq# in the put watch API.

Relates #37872 
Relates #10708

### CODE REVIEW

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java [hunk 1]
```diff
@@ -92,14 +94,18 @@ protected void doExecute(PutWatchRequest request, ActionListener<PutWatchRespons
 
                 if (isUpdate) {
                     UpdateRequest updateRequest = new UpdateRequest(Watch.INDEX, Watch.DOC_TYPE, request.getId());
-                    updateRequest.version(request.getVersion());
+                    if (request.getIfSeqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO) {
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java]: something missing here?

Comment (bleskes) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java]: ouch. yes.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java [hunk 2]
```diff
@@ -92,14 +94,20 @@ protected void doExecute(PutWatchRequest request, ActionListener<PutWatchRespons
...
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java]: why is this one not guarded by 

```
if (clusterService.state().nodes().getMinNodeVersion().onOrAfter(Version.V_7_0_0)) {
```

but the one in `TransportAckWatchAction` is?

**Comment by spinscale** [comment]
what about `TransportActivateWatchAction`? 

`TransportDeleteWatchAction` does not use versioning because we always want to delete. I think this is fine in this case as well?

**Review by spinscale — CHANGES_REQUESTED** [review]
few questions for clarification, but in general it looks good

**Comment by bleskes** [comment]
> what about TransportActivateWatchAction?

I don't see any versioning support there. Am I missing something?

> TransportDeleteWatchAction does not use versioning because we always want to delete. I think this is fine in this case as well?

That's what I was thinking.

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java (hunk 2 above)
Comment (bleskes) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java]: good catch

Comment (bleskes) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java]: this one let's the user decide which feature they use. The one in the ack watch action is internal and needs to make a decision on their own.

Comment (spinscale) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/transport/actions/put/TransportPutWatchAction.java]: makes sense

**Review by spinscale — APPROVED** [review]

**Review by jakelandis — APPROVED** [review]
Code LGTM.  However, I think we may want to revisit keeping support for the `version` since I don't believe Kibana is currently using it (I could be wrong, based a quick test) and I could not find documentation on it's usage.

**Comment by bleskes** [comment]
Thx @spinscale & @jakelandis.

> However, I think we may want to revisit keeping support for the version since I don't believe Kibana is currently using it 

The idea is to remove it indeed once I backport this all.

**Pull request merged**